src/Http/Http.Results/src/ResultsOfTHelper.cs

```diff
@@ -1,21 +1,53 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
 using System.Reflection;
+using System.Runtime.CompilerServices;
 using Microsoft.AspNetCore.Builder;
 using Microsoft.AspNetCore.Http.Metadata;
 
 namespace Microsoft.AspNetCore.Http;
 
 internal static class ResultsOfTHelper
 {
+    public const DynamicallyAccessedMemberTypes RequireMethods = DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods;
     private static readonly MethodInfo PopulateMetadataMethod = typeof(ResultsOfTHelper).GetMethod(nameof(PopulateMetadata), BindingFlags.Static | BindingFlags.NonPublic)!;
 
-    public static void PopulateMetadataIfTargetIsIEndpointMetadataProvider<TTarget>(MethodInfo method, EndpointBuilder builder)
+    // TODO: Improve calling static interface method with reflection
+    // https://github.com/dotnet/aspnetcore/issues/46267
+    public static void PopulateMetadataIfTargetIsIEndpointMetadataProvider<[DynamicallyAccessedMembers(RequireMethods)] TTarget>(MethodInfo method, EndpointBuilder builder)
     {
         if (typeof(IEndpointMetadataProvider).IsAssignableFrom(typeof(TTarget)))
         {
-            PopulateMetadataMethod.MakeGenericMethod(typeof(TTarget)).Invoke(null, new object[] { method, builder });
+            var parameters = new object[] { method, builder };
+
+            if (RuntimeFeature.IsDynamicCodeSupported)
+            {
+                InvokeGenericPopulateMetadata(parameters);
+            }
+            else
+            {
+                // Prioritize explicit implementation.
+                var populateMetadataMethod = typeof(TTarget).GetMethod("Microsoft.AspNetCore.Http.Metadata.IEndpointMetadataProvider.PopulateMetadata", BindingFlags.Static | BindingFlags.NonPublic);
+                if (populateMetadataMethod is null)
+                {
+                    populateMetadataMethod = typeof(TTarget).GetMethod("PopulateMetadata", BindingFlags.Static | BindingFlags.Public);
+                }
+                Debug.Assert(populateMetadataMethod != null, $"Couldn't find PopulateMetadata method on {typeof(TTarget)}.");
+
+                populateMetadataMethod.Invoke(null, BindingFlags.DoNotWrapExceptions, binder: null, parameters, culture: null);
+            }
+        }
+
+        // TODO: Remove IL3050 suppress when https://github.com/dotnet/linker/issues/2715 is complete.
+        [UnconditionalSuppressMessage("AOT", "IL3050", Justification = "Validated with IsDynamicCodeSupported check.")]
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2060:MakeGenericMethod",
+            Justification = "PopulateMetadataMethod calls a public static method. TTarget is annotated to include public methods.")]
```

Comment:
(nit) This justification doesn't match why the warning is happening. 

The reason the warning is happening is explained here:

https://github.com/dotnet/linker/blob/main/docs/error-codes.md#il2060-trim-analysis-call-to-systemreflectionmethodinfomakegenericmethod-can-not-be-statically-analyzed-its-not-possible-to-guarantee-the-availability-of-requirements-of-the-generic-method

It isn't that the `TTarget` might not have a public method on it. The reason is because the trimmer can't guarantee that 

```C#
private static void PopulateMetadata<TTarget>(MethodInfo method, EndpointBuilder builder) where TTarget : IEndpointMetadataProvider
```

Doesn't have a `[DynamicallyAccessMember]` attribute on its `<TTarget>` generic parameter. It can't statically tell which method `PopulateMetadataMethod` is pointing to. (or at least it doesn't today)

```suggestion
        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2060:MakeGenericMethod",
            Justification = "The call to MakeGenericMethod is safe due to the fact that  PopulateMetadata<TTarget> does not have a DynamicallyAccessMembers attribute.")]
```